// rg-quiz-app/vite.config.js
import { defineConfig } from 'vite';
import react from '@vitejs/plugin-react';
import { VitePWA } from 'vite-plugin-pwa';
import { visualizer } from 'rollup-plugin-visualizer';

export default defineConfig({
  base: '/',
  // --- ADDED: Server configuration for local development ---
  server: {
    host: '0.0.0.0', // Makes Vite listen on all network interfaces, which can help vercel dev detect it
    port: 5173,      // Explicitly sets the port for your frontend dev server
    proxy: {
      // This proxy is for when you run 'npm run dev' directly in the frontend folder.
      // When 'vercel dev' is running, it handles the /api proxying itself.
      // However, keeping it here ensures standalone frontend development works.
      '/api': {
        target: 'http://localhost:3000', // Default port for vercel dev's backend proxy
        changeOrigin: true,
        rewrite: (path) => path.replace(/^\/api/, '/api'), // Ensure the path remains /api/something
      },
    },
  },
  // --- END ADDED SERVER CONFIG ---
  build: {
    outDir: 'dist', // Explicitly set the output directory for the build
    chunkSizeWarningLimit: 1000,
    rollupOptions: {
      output: {
        manualChunks: {
          react: ['react', 'react-dom', 'react-router-dom'],
          ui: ['react-bootstrap', 'react-icons', 'lucide-react'],
          animation: ['framer-motion'],
          audio: ['howler'],
          fonts: ['@fortawesome/fontawesome-free'],
          pwa: ['vite-plugin-pwa']
        }
      },
      // --- ADDED: Externalize Node.js specific modules ---
      external: [
        'fsevents', // This was the primary error
        'path',
        'fs',
        'util',
        'stream',
        'os',
        'events',
        'process',
        'perf_hooks',
        'node:path',
        'node:fs',
        'node:fs/promises',
        'node:util',
        'node:stream',
        'node:os',
        'node:events',
        'node:process',
        'node:perf_hooks',
      ],
      // --- END ADDED SECTION ---
    }
  },
  plugins: [
    react(),
    VitePWA({
      registerType: 'autoUpdate',
      injectRegister: 'auto',
      workbox: {
        globPatterns: [
          '**/*.{js,css,html,ico,png,svg,webp,mp3,wav,json,webmanifest}',
        ],
        maximumFileSizeToCacheInBytes: 20 * 1024 * 1024,
        runtimeCaching: [
          {
            urlPattern: /\/all_quiz_data\.json/i,
            handler: 'StaleWhileRevalidate',
            options: {
              cacheName: 'quiz-data-cache',
              expiration: {
                maxEntries: 1,
                maxAgeSeconds: 60 * 60 * 24 * 30 // 30 days
              }
            }
          },
          {
            urlPattern: /\.(mp3|wav)$/i,
            handler: 'CacheFirst',
            options: {
              cacheName: 'audio-assets',
              expiration: {
                maxEntries: 50,
                maxAgeSeconds: 60 * 60 * 24 * 365 // 1 year
              }
            }
          },
          {
            urlPattern: /\.(png|svg|jpg|jpeg|webp)$/i,
            handler: 'CacheFirst',
            options: {
              cacheName: 'image-assets',
              expiration: {
                maxEntries: 100,
                maxAgeSeconds: 60 * 60 * 24 * 365 // 1 year
              }
            }
          }
        ],
        navigateFallback: '/index.html',
        clientsClaim: true,
        skipWaiting: true,
        cleanupOutdatedCaches: true,
        offlineGoogleAnalytics: false
      },
      devOptions: {
        enabled: true,
        type: 'module',
        navigateFallback: 'index.html'
      }
    }),
    visualizer({
      open: true,
      gzipSize: true,
      brotliSize: true,
      template: 'treemap'
    })
  ]
});
